src/Storage/WriteStream.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage;
+
+use Google\Cloud\Upload\AbstractUploader;
+use GuzzleHttp\Psr7\StreamDecoratorTrait;
+use GuzzleHttp\Psr7\BufferStream;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * A Stream implementation that uploads in chunks to a provided uploader when
+ * we reach a certain chunkSize. Upon `close`, we will upload the remaining chunk.
+ */
+class WriteStream implements StreamInterface
+{
+    use StreamDecoratorTrait;
+
+    private $uploader;
+    private $stream;
+    private $chunkSize = 262144;
```

Comment:
That sounds fair to me. What are your thoughts on allowing users to set the chunk size for the stream wrapper via `stream_context_create`?